includes/helper/oauth_helper.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+namespace Engelsystem\Events\Listener;
+
+use Engelsystem\Config\Config;
+use Engelsystem\Database\Database;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Collection;
+use Psr\Log\LoggerInterface;
+
+class OAuth2
+{
+    /** @var array */
+    protected $config;
+
+    /** @var LoggerInterface */
+    protected $log;
+
+    /**
+     * @param Config          $config
+     * @param LoggerInterface $log
+     */
+    public function __construct(Config $config, LoggerInterface $log)
+    {
+        $this->config = $config->get('oauth');
+        $this->log = $log;
+    }
+
+    /**
+     * @param string     $event
+     * @param string     $provider
+     * @param Collection $data OAuth userdata
+     */
+    public function login(string $event, string $provider, Collection $data)
+    {
+        /** @var Database $db */
+        $db = app(Database::class);
+        $ssoTeams = $this->getSsoTeams($provider);
+        $user = auth()->user();
+        $currentUserAngeltypes = collect($db->select('SELECT * FROM UserAngelTypes WHERE user_id = ?', [$user->id]));
+        $userGroups = $data->get(($this->config[$provider] ?? [])['groups'] ?? 'groups', []);
+
+        foreach ($userGroups as $groupName) {
+            if (!isset($ssoTeams[$groupName])) {
+                continue;
+            }
+
+            $team = $ssoTeams[$groupName];
+            $userAngeltype = $currentUserAngeltypes->where('angeltype_id', $team['id'])->first();
+            $supporter = $team['supporter'];
+            $arrived = $supporter ? $user->id : null;
+
+            if (!$userAngeltype) {
+                $this->log->info(
+                    'SSO {provider}: Added to angeltype {angeltype}, arrived: {arrived}, supporter: {supporter}',
+                    [
+                        'provider'  => $provider,
+                        'angeltype' => AngelType($team['id'])['name'],
+                        'arrived'   => $arrived ? 'yes' : 'no',
+                        'supporter' => $supporter ? 'yes' : 'no',
+                    ]
+                );
+                $db->insert(
+                    'INSERT INTO UserAngelTypes (user_id, angeltype_id, confirm_user_id, supporter) VALUES (?, ?, ?, ?)',
+                    [$user->id, $team['id'], $arrived, $supporter]
+                );
+
+                continue;
+            }
+
+            if (!$supporter) {
+                continue;
```

Comment:
When this is true, it's not possible to set confirm_user_id later. Maybe do confirm_user_id first?

Reply:
As you can only be confirmed automatically if you are a supporter thats imho ok that way?

Reply:
I see.